package/MDAnalysis/lib/nsgrid.pyx

```diff
@@ -0,0 +1,991 @@
+        searchgrid.fill_grid(searchcoords_bbox)
+        
+        
+        size_search = searchcoords.shape[0]
+
+        results = NSResults(self.cutoff, self.coords, searchcoords)
+
+        with nogil:
+            for i in range(size_search):
+                # Start with first search coordinate
+                current_beadid = i
+                # find the cellindex of the coordinate
+                cellindex = searchgrid.cellids[current_beadid]
+                for xi in range(DIM):
+                    for yi in range(DIM):
+                        for zi in range(DIM):
+                            check = True
+                            #Probe the search coordinates in a brick shaped box
+                            probe[XX] = searchcoords_bbox[current_beadid, XX] + (xi - 1) * searchgrid.cellsize[XX]
+                            probe[YY] = searchcoords_bbox[current_beadid, YY] + (yi - 1) * searchgrid.cellsize[YY]
+                            probe[ZZ] = searchcoords_bbox[current_beadid, ZZ] + (zi - 1) * searchgrid.cellsize[ZZ]
+                            # Make sure the probe coordinates is inside the brick-shaped box
+                            if self.periodic:
+                                for m in range(DIM - 1, -1, -1):
+                                    while probe[m] < 0:
+                                        for d in range(m+1):
+                                            probe[d] += self.box.c_pbcbox.box[m][d]
+                                    while probe[m] >= self.box.c_pbcbox.box[m][m]:
+                                        for d in range(m+1):
+                                            probe[d] -= self.box.c_pbcbox.box[m][d]
+                            else:
+                                for m in range(DIM -1, -1, -1):
+                                    if probe[m] < 0:
+                                        check = False
+                                        break
+                                    if probe[m] > self.box.c_pbcbox.box[m][m]:
+                                        check = False
+                                        break
+                            if not check:
+                                continue
+                            # Get the cell index corresponding to the probe
+                            cellindex_probe = self.grid.coord2cellid(probe)
+                            #for this cellindex search in grid
+                            for j in range(self.grid.nbeads[cellindex_probe]):
+                                bid = self.grid.beadids[cellindex_probe * self.grid.nbeads_per_cell + j]
+                                #find distance between search coords[i] and coords[bid]
+                                d2 = self.box.fast_distance2(&searchcoords_bbox[current_beadid, XX], &self.coords_bbox[bid, XX])
+                                if d2 < cutoff2 and d2 > EPSILON:
+                                    results.add_neighbors(current_beadid, bid, d2)

```

Comment:
Most of the capped funcctions have a return signature such that same atom in different group (query and reference coordinates) will be obtained as a neighbour. In that case, should I remove ``EPSILON`` here. It will be included only in ``self_search`` function.

Reply:
Yeah remove that epsilon hack so this works as other distance methods